Create a non-global Client

shepmaster opened the pull request:
Only moving the bucket access control functions right now.

I wanted to open this early to get some feedback from you on how you feel about this API. I'm leaving the existing functions / globals behind in order to do this piecemeal, but I'd expect you could remove those immediately afterward or after an intermediate release to allow people to upgrade.

I'll continue on this path in the hopes that it looks acceptable, but please let me know if you have any concerns or questions!

ThouCheese commented:
Hi Jake, thanks for this PR, would this resolve issue #60 is it were finished? I see that you have moved all functions into a single file, and then call those functions from the respective resource with the global client. The drawbacks are that now that file becomes fairly large, and secondly all functions that do not use a global client are documented in a single webpage, which becomes a bit unwieldy because of this. I don't see a solution without some drawbacks of its own however, so I think this approach is fine.

shepmaster commented:
> Hi Jake, thanks for this PR, would this resolve issue #60 is it were finished?

That is the goal, yes.

> I see that you have moved all functions into a single file, and then call those functions from the respective resource with the global client.

Yes. I hope to provide a feature flag that would disable the global client usage as well, helping guide new users to the better path.

> The drawbacks are that now that file becomes fairly large, and secondly all functions that do not use a global client are documented in a single webpage, which becomes a bit unwieldy because of this. 

Indeed. You'll see the same pattern used in similar crates, [such as Rusoto](https://docs.rs/rusoto_ec2/0.46.0/rusoto_ec2/trait.Ec2.html).

> I don't see a solution without some drawbacks of its own however, so I think this approach is fine.

The other option I know is to have "view objects" that hang off the initial client:

```rust
struct Client;

impl Client {
    fn thing_one(&self) -> ThingOne<'_> { todo!() }
    fn thing_two(&self) -> ThingTwo<'_> { todo!() }
}

struct ThingOne<'a>(&'a Client);

impl<'a> ThingOne<'a> {
    async fn create(&self) { todo!() }
}

struct ThingTwo<'a>(&'a Client);

impl<'a> ThingTwo<'a> {
    async fn create(&self) { todo!() }
}
```

ThouCheese commented:
I have given this some thought and played around a bit with the documentation and I really like the view object approach, because it makes the resulting documentation more manageable. (Also, not as relevant, but they are completely optimized out as far as I can tell from the disassembler, so :partying_face: ). I feel like that is the best way forward.

shepmaster commented:
> I feel like that is the best way forward.

Sounds good! 

What kind of git history would you like to have? What I have right now is one commit moves each section at a time to the big client. I'd be happy to go commit-by-commit and redo those commits to move to a separate view objects instead. If you prefer having one big-bang commit, I'll just pile on more changes and squash them all at the end.

ThouCheese commented:
I don't mind having a single large squashed commit

shepmaster commented:
Alright, I've...

- rebased on master, incorporating the recent changes
- moved everything to view objects
- updated the related doctests
- added a new feature flag that keeps the global client functions around

Some next steps:

1. Do we wish to continue to have synchronous adapters? If so, I think we should follow reqwest's pattern and have `sync::client::Client` / `sync::client::BucketClient`.
    - Should the documentation for these point to the async version, like the current ones, or do you want me to copy-pasta and update the async docs?
1. Speaking of docs, should I reduce the docs on the global client methods to point to the non-global counterparts?
1. I haven't written any docs for the _new_ things (`Client` / `BucketClient` / `Client::bucket`, etc.). You have any guidance for what you'd like to see on those?

ThouCheese commented on src/lib.rs:
````diff
@@ -83,13 +83,15 @@
 //! ```
 #![forbid(unsafe_code, missing_docs)]
 
+pub mod client;
````
Should this be a public module?

shepmaster replied:
It depends on how you wish to expose the sub-clients. They are `crate::client::SpecificClient` right now. If `mod client` is not public, you can not click through to their documentation:

<img width="1053" alt="image" src="https://user-images.githubusercontent.com/174509/112691335-89303600-8e53-11eb-9453-f367a476eed1.png">

ThouCheese replied:
Right fair enough! Lets leave this the way it is then

ThouCheese reviewed (commented):
I have one small question.

ThouCheese commented:
The progress so far seems good! To answer your questions:

1. Agree that that following `Reqwest`s pattern here is the best option. If we do that I think that the doc comments should be copy pasted, since they will no longer be adjacent to the method that they are referencing.
2. I think leaving the global method's comments as-is is best here. (With this level of duplication there should be a way to indicate that a function should copy another function's doc comment :thinking:)
3. I don't have strong opinions on how docs should be written, just a best effort attempt is fine by me.

shepmaster commented:
All right, I _believe_ I have all the pieces in play now.

ThouCheese commented on src/token.rs:
```diff
@@ -34,29 +32,29 @@ impl Token {
-    async fn get_token(scope: &str) -> Result<(String, u64), Error> {
-        let now = now();
+    async fn get_token(client: &reqwest::Client, scope: &str) -> crate::Result<(String, u64)> {
+        let now = crate::token::now();
```
Why are these paths fully specified, instead of just writing `now()`?

shepmaster replied:
Huh, very strange. I might have moved some code around (needing to qualify it) and then moved it back. Fixed.

ThouCheese replied:
Rust-analyzer to the rescue I'm sure :P

ThouCheese reviewed (changes requested):
I have two small remarks, but other than that this looks like an amazing improvement!

ThouCheese commented on src/client.rs:
```diff
@@ -0,0 +1,81 @@
+    }
+}
+
+impl Client {
```
It might be sensible to include a function in `Client` called `new`, that just delegates to `Client::default`, because people will expect to be able to call `new` and get a new `Client`

shepmaster replied:
Yep, I debated that. I tend to just to `default` in my _own_ code because I'm lazy ;-). Fixed.

ThouCheese replied:
Brilliant

nhynes commented:
I made another GCS crate that you may consider perusing for inspiration: https://crates.io/crates/cloud-storage-lite

shepmaster commented:
Changes [force pushed](https://github.com/ThouCheese/cloud-storage-rs/compare/c173aca5ea8ac744c5e3420d9357f2a995df500d..ee46704c26fc6e00f9f0eb0fd268f91b7b781e02)

ThouCheese commented:
LGTM

ThouCheese commented:
I'm merging this pull request into master. I will release a new version of cloud-storage in a day or so, so you can expect version `0.9` soon :)

ThouCheese commented:
A day or so got stretched a bit, but we are now on 0.9! It includes the debug impl as well.
